### PR TITLE
Add support for building portlets.

### PR DESCRIPTION
**Pull request by mbaechtold**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -486,6 +486,19 @@ The ZCML builder builds a ZCML file:
...
+The ``ftw.builder`` ships with a few builders for Plone portlets, but the
+idea is that you can easily craft your own builders for your portlets or
+extend existing builders.
+
+The built-in builders are:
+
+- ``static portlet`` - creates an static portlet
+- ``navigation portlet`` - creates an navigation portlet
+
+
```
Comment (jone) [inline README.rst]: Could you add an example how I can easily craft my own builder? :wink:

File: ftw/builder/portlets.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.before_create()
+        manager, assignments = self.get_manager_and_assignments()
+        portlet = self.create_portlet(assignments)
+        self.after_create(manager, assignments, portlet)
+        return portlet
+
+    def create_portlet(self, assignments):
+        portlet = self.assignment_class(**self.arguments)
+        name = self.choose_name(assignments, portlet)
+        portlet.__name__ = portlet
```
Comment (jone) [inline ftw/builder/portlets.py]: this looks wrong, shouldn't that be
```python
portlet.__name__ = name
```
:question:

File: ftw/builder/tests/__init__.py [hunk 3]
```diff
@@ -14,3 +15,13 @@ def setUp(self):
...
+
+
+class FunctionalTestCase(TestCase):
+
+    layer = BUILDER_FUNCTIONAL_TESTING
+
+    def setUp(self):
+        self.portal = self.layer['portal']
+        setRoles(self.portal, TEST_USER_ID, ['Manager'])
+        login(self.portal, TEST_USER_NAME)
```
Comment (jone) [inline ftw/builder/tests/__init__.py]: login should not be necessary - the test user should already be logged in. that's a common mistage :wink:

File: ftw/builder/tests/__init__.py [hunk 4]
```diff
@@ -14,3 +15,13 @@ def setUp(self):
...
         login(self.portal, TEST_USER_NAME)
+
+
+class FunctionalTestCase(TestCase):
+
+    layer = BUILDER_FUNCTIONAL_TESTING
+
+    def setUp(self):
+        self.portal = self.layer['portal']
+        setRoles(self.portal, TEST_USER_ID, ['Manager'])
```
Comment (jone) [inline ftw/builder/tests/__init__.py]: I'd prefer a `grant` method here, e.g.:
```python
    def grant(self, *roles):
        setRoles(self.portal, TEST_USER_ID, list(roles))
        transaction.commit()
```

and then grant the role in each test where it is required, e.g.:

```python
    def test_static_portlet_found_in_leftcolumn(self, browser):
        self.grant('Manager')
        ...
```

The advantage is that we do not do it when it is not necessary and that it is explicitly declared in the test that we have a different role for the test user. It makes reading the test easier. :wink:

**Comment by mbaechtold** [comment]
Thx @jone for the feedback. Could you please have another look?

File: README.rst [hunk 5]
```diff
@@ -486,6 +486,31 @@ The ZCML builder builds a ZCML file:
...
+    class MyPortletBuilder(PlonePortletBuilder):
+        assignment_class = my_portlet.Assignment
+
+    builder_registry.register('my portlet', MyPortletBuilder)
+
+
+The built-in builders are:
+
+- ``static portlet`` - creates an static portlet
+- ``navigation portlet`` - creates an navigation portlet
```
Comment (jone) [inline README.rst]: its `a portlet`, not `an portlet`, I think :wink:

File: README.rst [hunk 6]
```diff
@@ -486,6 +486,31 @@ The ZCML builder builds a ZCML file:
...
+
+The ``ftw.builder`` ships with a few builders for Plone portlets, but the
+idea is that you can easily craft your own builders for your portlets or
+extend existing builders.
+
+Example:
+
+.. code:: python
+
+    from ftw.builder.portlets import PlonePortletBuilder
```
Comment (jone) [inline README.rst]: it might be helpful if you include the ``builder_registry`` import here, just for copy/paste :wink:

**Comment by mbaechtold** [comment]
@jone: updated :smiley:

**Comment by jone** [comment]
cool :+1:

**Pull request merged**